### PR TITLE
Add documentation for Handshake dialect

### PR DESCRIPTION
**Pull request by xgupta**
This PR intended to add documentation for the handshake dialect and handshake-runner.

### CODE REVIEW

**Review by stephenneuendorffer — CHANGES_REQUESTED** [review]
Thanks for starting this!

File: docs/CommandGuide/handshake-runner.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+% handshake-runner(1) | "handshake-runner" Documentation
```
Comment (lattner) [inline docs/CommandGuide/handshake-runner.md:1]: Nice.  What is this first line?

File: docs/CommandGuide/handshake-runner.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+## NAME
+
+**handshake-runner** — Run restricted std MLIR and CIRCT handshake IR!
```
Comment (lattner) [inline docs/CommandGuide/handshake-runner.md:5]: I'd recommend dropping the `!` at the end

**Review by lattner — COMMENTED** [review]
Thank you for working on this!

File: docs/CommandGuide/handshake-runner.md (hunk 1 above)
Comment (xgupta) [reply docs/CommandGuide/handshake-runner.md:1]: It is a metadata block for pandoc. I use an [internet resource](https://eddieantonio.ca/blog/2015/12/18/authoring-manpages-in-markdown-with-pandoc/) to write a markdown to manpage convertible document using pandoc. I think handshake-runner doesn't need to have a manpage so should drop this line and the next `NAME`  line.

File: docs/Dialects/Handshake.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+Operation documentation is written in each op’s Op Definition Spec using 
+TableGen. A markdown version of the doc can be generated using `mlir-tblgen 
+-gen-op-doc` and is attached in the [Operation definitions](#operation-definitions) 
+section.
+
+## Interfaces
+
+MergeLikeOpInterface - Returns an operand range over the data signals being merged.
+Useful verification purposes during lowering from other dialects.
```
[the remark this answers is not in the document]
Comment (xgupta) [reply docs/Dialects/Handshake.md:80]: Yes, Tablegen should produce md for fooInterfaces.td. We can directly include them from there. 
But currently, I think we don't have that set up in CMakelist.txt. 
I am removing things that can be generated by Tablgen now and will add the rule for these in CMakelist.txt.

File: docs/Dialects/Handshake.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+The interface declaration is auto-generated from TableGen definitions 
+included in HandshakeOps.td .
```
[the remark this answers is not in the document]
Comment (xgupta) [reply docs/Dialects/Handshake.md:83]: removed

File: docs/Dialects/Handshake.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+## Conversions
+
+HandshakeBuilder - Handshake Builder class \
+HandshakeFuncOpLowering - HandshakeTofoo lowering class
```
[the remark this answers is not in the document]
Comment (xgupta) [reply docs/Dialects/Handshake.md:88]: removed

File: docs/Dialects/Handshake.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+### The dialect 
+
+The code for the Handshake dialect resides in a few places-
+
+  Public headers are placed in include/circt/Dialect/Handshake . \
+  Libraries are placed in lib/Dialect/Handshake . \
+  IR tests are placed in test/Dialect/Handshake . 
+
+The dialect itself, including all types and ops, is in the CIRCTHandshakeOps 
+library.
```
[the remark this answers is not in the document]
Comment (xgupta) [reply docs/Dialects/Handshake.md:101]: removed

File: docs/Dialects/Handshake.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+## Operations
+
+In Handshake, instruction is a generalized concept; a Handshake function is just
+a sequence of instructions. Declaring types, expressing computations, annotating 
+result ids, expressing control flows and others are all in the form of 
+instructions.
+
+Operation documentation is written in each op’s Op Definition Spec using 
+TableGen. A markdown version of the doc can be generated using `mlir-tblgen
```
[the remark this answers is not in the document]
Comment (xgupta) [reply docs/Dialects/Handshake.md:73]: removed

**Comment by xgupta** [comment]
Thanks for reviewing! The intention was to start some documentation. I will update it after gaining more experience. I think some concepts of handshake region, process, FIFO need some light.

**Comment by xgupta** [comment]
Two small questions-
1. I named folder `docs/Dialects` instead of `docs/Dialect` because MLIR uses the former. but when I see  `add_circt_doc` destination folder named as `/Dialect` . Which one should I choose?
2. When I see the` add_circt_doc(HandshakeOps -gen-op-doc handshake Dialect/)` the output file is named as `handshake` while the given command option is `-gen-op-doc`, should handshake renamed as `handshakeops` following the MLIR conventions?

File: docs/CommandGuide/handshake-runner.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+```
+module {
+  func @main(%arg0: index) -> (i8) {
+    %0 = alloc() : memref<10xi8>
+    %c1 = constant 1 : i8
+         store %c1, %0[%arg0] : memref<10xi8>
+         %1 = load %0[%arg0] : memref<10xi8>
+         return %1 : i8
+  }
+}
```
Comment (hanchenye) [inline docs/CommandGuide/handshake-runner.md:29]: The indentation is a little weird on my side.

File: docs/Dialects/Handshake.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+The Handshake dialect adopts the following conventions for IR:
+- The prefix for all Handshake types and operations are `handshake.`.
+
+## Resources
+
+MLIR Handshake Dialect-[slides](https://drive.google.com/file/d/1UYQAfHrzcsdXUZ93bHPTPNwrscwx89M-/view?usp=sharing) by Stephen Neuendorffer (Xilinx) + Lana Josipović (EPFL)
+
+## Operation definitions
+
+[include "Dialects/HandshakeOps.md"]
```
Comment (hanchenye) [inline docs/Dialects/Handshake.md:64]: Would it make sense to add a link pointing to the source file?

**Review by hanchenye — APPROVED** [review]
Thanks for creating this!

File: docs/CommandGuide/handshake-runner.md (hunk 8 above)
Comment (xgupta) [reply docs/CommandGuide/handshake-runner.md:29]: It is picked from a testcase see [handshake-runner/loadstore.mlir](https://github.com/llvm/circt/blob/main/test/handshake-runner/loadstore.mlir.).

File: docs/Dialects/Handshake.md (hunk 9 above)
Comment (xgupta) [reply docs/Dialects/Handshake.md:64]: It is something the same as MLIR's dialect documentation for example [see GPU](https://github.com/llvm/llvm-project/blob/main/mlir/docs/Dialects/GPU.md#operations) and operational state in [MLIR website](https://mlir.llvm.org/docs/Dialects/GPU/#operations). HandshakeOps.md will be generated from [HandshakeOps.td](https://github.com/llvm/circt/blob/main/include/circt/Dialect/Handshake/HandshakeOps.td) using mlir-tblgen.

Comment (hanchenye) [reply docs/Dialects/Handshake.md:64]: Auh I see, thanks for letting me know.

**Comment by xgupta** [comment]
I think I should merge this PR as I got one approval. And we will make further improvements in an iterative way.

**Pull request merged**